File: lib/image/ExtractCO.dart
import 'dart:typed_data';

import 'package:json_annotation/json_annotation.dart';
import 'package:raising/common/JsonConverter.dart';


part 'ExtractCO.g.dart';

@JsonSerializable()
@CustomDateTimeConverter()
@MapIntUint8listConverter()
class ExtractCO {
  String msg;
  String error;

  /**
   * 在webdav中，这个字段会是uri编码的，因为解码再编码，无法处理（号，所以需要这样做，
   * 之后webdav 数据库搜索（保存了webdav 编码后的absPath），需要处理这种情况
   * 为什么要这样处理？因为webdav的行为看起来很难评估，所以这部分的逻辑真不应该乱搞。
   * 如果存两个absPath 跟orignalAbsPath，也难解决这个问题，因为还是无法解决存到数据库的absPath应该是哪个的问题。
   *
   * 说到底，我当初考虑absPath的时候，就没考虑处理这种要转码的情况的语义（也不知道absPath的相关功能）
   * 目前定义为host相关的唯一原生标识即可，需要转成别的形式的话到时再处理，这样比较好，先用比较原始的语义处理。
   */
  String absPath; //总是包括filename(如果是文件夹，则包括到文件夹名），只有当没有absPath时，再去查filename字段
  String filename;
  int size;
  int fileNum; //如果是压缩文件里面有多少文件，如果是目录，里面有多少文件
  bool isDirectory = false;
  bool isCompressFile = false;
  DateTime createTime;
  DateTime updateTime;
  String compressFormat;

  //index 按文件名（包括路径排序，第几个文件，跳过文件夹）的排序序号，从0开始。
  Map<int, String> indexPath; //压缩文件内的绝对路径
  Map<int, Uint8List> indexContent;

  factory ExtractCO.fromJson(Map<String, dynamic> json) => _$ExtractCOFromJson(json);

  Map<String, dynamic> toJson() => _$ExtractCOToJson(this);

  ExtractCO();
}

File: android/app/src/main/kotlin/com/example/raising/vo/ExtractCO.kt
package com.example.raising.vo

import java.util.*
import kotlin.collections.HashMap

/**
 * 我的设计又犯了错误，由于extract 网络文件的信息，很有可能能获取到原本不需要的信息，比如压缩文件里的名字，但这些信息又可能在某个地方有用
 * 即使是在解压他们的内容也会有这样的问题，为了不浪费时间再重传一次，
 * 设计的ExtractCO应该不再严格保证能获取到什么样的内容，而是能获取到什么，就塞什么进去，当然，字段的命名要有规范。
 * 另外，由于压缩文件的属性是有限的，所以这种做法应该没什么问题。
 */
class ExtractCO {
    var msg: String = "OK";
    var error: String = "";
    var absPath: String? = null;//总是包括filename(如果是文件夹，则包括到文件夹名），只有当没有absPath时，再去查filename字段
    var filename: String? = null;
    var size: Long? = null;
    var fileNum: Int? = null;//如果是压缩文件里面有多少文件，如果是目录，里面有多少文件
    var isDirectory = false
    var isCompressFile = false
    var createTime: Date? = null
    var updateTime: Date? = null
    var compressFormat: String? = null;

    //index 按文件名（包括路径排序，第几个文件，跳过文件夹）的排序序号，从0开始。
    var indexPath: HashMap<Int, String>? = null;//压缩文件内的绝对路径
    var indexContent: HashMap<Int, ByteArray>? = null;

    fun getMap(): HashMap<String, Any> {
        val cur = HashMap<String, Any?>();

        cur["msg"] = msg;
        cur["error"] = error;
        cur["absPath"] = absPath
        cur["filename"] = filename
        cur["size"] = size;
        cur["fileNum"] = fileNum;
        cur["isDirectory"] = isDirectory;
        cur["isCompressFile"] = isCompressFile;
        cur["createTime"] = createTime?.time;
        cur["updateTime"] = updateTime?.time;
        cur["compressFormat"] = compressFormat;
        cur["indexPath"] = indexPath;
        cur["indexContent"] = indexContent;

        val res = HashMap<String, Any>();
        cur.forEach {
            if (it.value != null) {
                res[it.key] = it.value!!
            }
        }
        return res;
    }

}
